src/jira/util/jwt.ts

```diff
@@ -2,50 +2,56 @@
 // https://bitbucket.org/atlassian/atlassian-connect-express/src/f434e5a9379a41213acf53b9c2689ce5eec55e21/lib/middleware/authentication.js?at=master&fileviewer=file-view-default#authentication.js-227
 // TODO: need some typing for jwt
 import jwt from "atlassian-jwt";
-import { Request, Response } from "express";
-
-const TOKEN_KEY_PARAM = "acpt";
-const TOKEN_KEY_HEADER = `X-${TOKEN_KEY_PARAM}`;
+import {Request, Response} from "express";
+import envVars from "../../config/env";
 
 const JWT_PARAM = "jwt";
 const AUTH_HEADER = "authorization"; // the header name appears as lower-case
+const BASE_URL = envVars.APP_URL
+
+/**
+ * Atlassian Connect has 2 different types of JWT tokens.
+ * Normal tokens has qsh parameter and generated by Jira for backend authentication (like installation webhooks)
+ * or web elements like iFrame
+ *
+ * Context tokens are tokens which are generated by App iframes for the authentication with app backend.
+ * They don't require sqh verification and their qsh is set to a fixed `context-qsh` value.
+ *
+ * When building endpoints we should specify which type of tokens they require
+ *
+ * See details at: https://community.developer.atlassian.com/t/action-required-atlassian-connect-vulnerability-allows-bypass-of-app-qsh-verification-via-context-jwts/47072
+ */
+export enum TokenType {
+	normal = "normal",
+	context = "context"
+}
 
 function extractJwtFromRequest(req) {
-	const tokenInQuery = req.query[JWT_PARAM];
-
-	// JWT is missing in query and we don't have a valid body.
-	if (!tokenInQuery && !req.body) {
-		req.log(
-			`${"Cannot find JWT token in query parameters. " +
-			"Please include body-parser middleware and parse the urlencoded body " +
-			"(See https://github.com/expressjs/body-parser) if the add-on is rendering in POST mode. " +
-			"Otherwise please ensure the "}${JWT_PARAM} parameter is presented in query.`
-		);
-		return;
-	}
+	const tokenInQuery = req.query ? req.query[JWT_PARAM] : undefined;
 
 	// JWT appears in both parameter and body will result query hash being invalid.
-	const tokenInBody = req.body[JWT_PARAM];
+	const tokenInBody = req.body ? req.body[JWT_PARAM] : undefined;
 	if (tokenInQuery && tokenInBody) {
 		req.log("JWT token can only appear in either query parameter or request body.");
 		return;
 	}
 	let token = tokenInQuery || tokenInBody;
 
 	// if there was no token in the query-string then fall back to checking the Authorization header
-	const authHeader = req.headers[AUTH_HEADER];
+	const authHeader = req.headers ? req.headers[AUTH_HEADER] : undefined;
 	if (authHeader?.startsWith("JWT ")) {
 		if (token) {
 			const foundIn = tokenInQuery ? "query" : "request body";
-			req.log(`JWT token found in ${foundIn} and in header: using ${foundIn} value.`);
+			req.log.info(`JWT token found in ${foundIn} and in header: using ${foundIn} value.`);
 		} else {
 			token = authHeader.substring(4);
 		}
 	}
 
-	// TODO: Remove when we discontinue the old token middleware
```

Comment:
(View as side-by-side diff)
I removed this hence this part of the code already was unreachable and we don't use these parameter names for our tokens